### ISSUE TITLE
PSD file ImageIO.read gives java.io.EOFException: Unexpected end of PackBits stream

### ISSUE THREAD

**Issue by prakharmishra** [haraldk/TwelveMonkeys#179]
Hey everyone,

I am trying to convert a PSD file (500+ MB in size) which yields to this exception:
```java
java.io.EOFException: Unexpected end of PackBits stream
	at com.twelvemonkeys.io.enc.PackBitsDecoder.readFully(Unknown Source)
	at com.twelvemonkeys.io.enc.PackBitsDecoder.decode(Unknown Source)
	at com.twelvemonkeys.io.enc.DecoderStream.fill(Unknown Source)
	at com.twelvemonkeys.io.enc.DecoderStream.read(Unknown Source)
	at java.io.DataInputStream.readShort(DataInputStream.java:313)
	at com.twelvemonkeys.imageio.plugins.psd.PSDImageReader.read16bitChannel(Unknown Source)
	at com.twelvemonkeys.imageio.plugins.psd.PSDImageReader.readImageData(Unknown Source)
	at com.twelvemonkeys.imageio.plugins.psd.PSDImageReader.read(Unknown Source)
	at javax.imageio.ImageIO.read(ImageIO.java:1448)
	at javax.imageio.ImageIO.read(ImageIO.java:1308)
```

Could anybody please suggest a solution for this? Or, at least, a reason why it yields this exception (maybe newer version of photoshop?).

**UPDATE:**

I am using Java 8 with TwelveMonkeys 3.1.1 and here is the link for the PSD file I used:
[heavy.psd][1]

  [1]: https://drive.google.com/file/d/0Bx8GyTBfuSFuZEgwRl9vYXdLX1k/view?usp=sharing

Thank you

**Comment by haraldk** [comment]
Hi,

Thanks for reporting!

This could be a bug, but it's hard to say without being able to reproduce. Could also be a corrupt PSD file, or as you say, a new version of Photoshop (although unlikely, as the PackBits algorithm is well-known).
 
Technically, the exception happens because the decoder runs out of data, before it has fully read all the 16 bit samples for a channel. Most test PSD files that I have, use 8 bits per channel, so you might have run into an issue.

Can you please share the file you are trying to read, so that I can have a look at it, and fix it if it turns out its a bug?

It's possible to attach files to this issue, by renaming `file.psd` to `file_psd.png` and drop it into the comments box. But as the file is quite large, it's perhaps better to share via Google Drive, Dropbox or similar file sharing service.

Also, please state which version of Java and TwelveMonkeys you are using.


Thanks,

-- 
Harald K

**Comment by haraldk** [comment]
See also related issue on StackOverflow: http://stackoverflow.com/q/32838766/1428606

**Comment by haraldk** [comment]
Okay, so here's the issue:

The PSD in question is 16 bit per sample, and uses PackBits compression. For some reason, I have had no test samples with this specific configuration, so there's been a bug there for ages. 

For PackBits-compressed PSDs, there's a list of each row's byte count (for easier random access I guess). I had this value multiplied by 2 for 16 bit and 4 for 32 bit samples, but obviously a byte count is just a byte count, regardless of if the samples are 2 or 4 bytes...

Now, when I changed this, the image is read without error. However, this PSD seems to be stored without a "merged" layer, yet it has 7 channels, of which 4 are alpha channels... And no layers..? When I open it, it just comes out mostly transparent (opening in Preview, it comes out all transparent).


Regards,

-- 
Harald K

**Comment by prakharmishra** [comment]
That's great. 

I envy your in-depth knowledge of **Image Formats** and their compression algorithms, Layered structure, etc. Wish I could have helped.

Actually, in photoshop, the image is not transparent at all. This is how it opens in Photoshop (CS6 Extended, at least):

![scenary](https://cloud.githubusercontent.com/assets/6131734/10215070/19e0faca-683c-11e5-9600-d7878e66a483.png)

I don't even know why this file is ridiculously heavy. Never mind that.

**Comment by haraldk** [comment]
Short update,

Did some work on this last week, and implemented reading of some of the extra layer data, but I still can't reconstruct the image above...

The reason for the file being heavy, is just.. It's PSD. ;-)

No, it's large because it's 16 bits per sample, and contains a lot of transparency masks and (seemingly redundant) layer data in addition to a lot of extra meta data. You could probably shave off about 3/4 of the file, without losing any information (pixel data, that is).

-- 
Harald K

**Comment by haraldk** [comment]
Fixed EOFExceptions in b85d0f7 .

-- 
Harald K
